### PR TITLE
Add genomes on server functionality

### PR DESCRIPTION
**Pull request by elischberg**

### CODE REVIEW

**Comment by peterjc** [comment]
Can you point me at the Galaxy tool author documentation for this new (to me) functionality? I'm wondering e.g. if it requires a specific version of Galaxy (and if it would fail gracefully on an older one).

**Comment by wm75** [comment]
@peterjc https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-inputs-param-options
not sure when this was added, but definitely long ago (before 2015 I'd say) :-)

File: tools/ncbi_blast_plus/ncbi_makeblastdb.xml [hunk 1]
```diff
@@ -60,52 +65,64 @@ $hash_index
...
+        <option value="nucl">nucleotide</option>
+      </param>
+      <!-- TODO Allow merging of existing BLAST databases (conditional on the database type)?
+           NOTE Double check the new database would be self contained first
+      -->
+      <!-- Note this is a mandatory parameter - default should be most recent FASTA file -->
+      <conditional name="input_selection">
+        <param name="source" type="select" label="Input is a">
+          <option value="history">Dataset in history</option>
+          <option value="cached">Genome on server</option>
```
Comment (wm75) [inline tools/ncbi_blast_plus/ncbi_makeblastdb.xml:79]: You need another conditional level here. Selecting a "Genome on server" makes no sense when the user wants to build a protein database.

Comment (peterjc) [reply tools/ncbi_blast_plus/ncbi_makeblastdb.xml:79]: Good point. Maybe expand the existing choice:

```xml
     <param argument="-dbtype" type="select" display="radio" label="Molecule type of input">
        <option value="prot">protein</option>
        <option value="nucl">nucleotide</option>
      </param>
```

to something like this:

```xml
     <param argument="-dbtype" type="select" display="radio" label="Molecule type of input">
        <option value="prot">protein from your history</option>
        <option value="nucl">nucleotide from your history</option>
        <option value="nucl-server">nucleotide from server</option>
      </param>
```

Comment (elischberg) [reply tools/ncbi_blast_plus/ncbi_makeblastdb.xml:79]: True! (:

Comment (wm75) [reply tools/ncbi_blast_plus/ncbi_makeblastdb.xml:79]: Would be the simplest way, but on second thought has the downside of breaking the 1:1 relationship between `$dbtype` and the command line option. Nested conditionals are slightly more complex, but the cleaner solution.

**Comment by elischberg** [comment]
I had problems in handling the term "value_label" in the output name, because i didn't understand it quite well. Is it only for having "Protein" instead of "prot" in the name of the output?
I removed it for now, but I can insert it again, if I get to know how to use it.

**Comment by peterjc** [comment]
Would the current nested conditionals break old workflows etc? Might therefore this be a safer approach:

```xml
     <param argument="-dbtype" type="select" display="radio" label="Molecule type of input">
        <option value="prot">protein from your history</option>
        <option value="nucl">nucleotide from your history</option>
        <option value="nucl-server">nucleotide from server</option>
      </param>
```

File: tools/ncbi_blast_plus/ncbi_makeblastdb.xml [hunk 2]
```diff
@@ -36,7 +51,12 @@ $hash_index
 ##Would default to being based on the cryptic Galaxy filenames, which is unhelpful
 -title 'BLAST Database'
 #end if
--dbtype $dbtype
+-dbtypne
```
Comment (wm75) [inline tools/ncbi_blast_plus/ncbi_makeblastdb.xml:54]:
```suggestion
-dbtype
```

**Review by wm75 — COMMENTED** [review]
Nice :) Thanks for adding the new test!

You will need to bump the @VERSION_SUFFIX@ eventually.

**Comment by wm75** [comment]
> Would the current nested conditionals break old workflows etc?

@peterjc in a sense, yes. Though of course only for people that are trying to run their workflow against a server that doesn't have the old tool version (requested by the workflow) installed. We normally consider this an acceptable drawback though I don't like making the life of users unnecessarily hard either.
In this case, however, there is no way to add the feature without this kind of breakage. Even your proposed simpler solution would require a new outside conditional to treat the new third selection differently, and again that would make workflow auto-updates to the new version impossible.

After testing a few alternatives I suggested to @elischberg to use the repeat elements (instead of the original multi-selects) because that lets you combine FASTAs from the history and server-provided ones when building the DB.
Was hoping to be able to propose a smaller change, but it didn't work out. Sorry for the extra reviewing effort and hope you find this solution acceptable.

**Comment by wm75** [comment]
btw, the use case behind the new feature is a viral primer scheme design tool that lets you exclude primer candidates based on them having matches in a custom BLAST DB, e.g. to avoid amplification of host DNA. With the new option, users will be able to create their desired subtraction DB as long as Galaxy Europe has the corresponding host reference installed.

**Comment by peterjc** [comment]
Given both @elischberg and @wm75 have been testing this (I have not), I'm happy in principle to merge this.

There are not any other imminent pull requests pending, so I suggest you bump the version and describe the changes in https://github.com/peterjc/galaxy_blast/blob/master/tools/ncbi_blast_plus/README.rst - something like:

```
2.14.1+galaxy1 - Use genome FASTA files on the server with ``makeblastdb`` (contribution from ...)
```

**Comment by peterjc** [comment]
Test failures in GitHub actions:

```
 "failure;ncbi_makeblastdb;0"
"failure;ncbi_makeblastdb;1"
"failure;ncbi_makeblastdb;2"
"failure;ncbi_makeblastdb;3"
"failure;ncbi_makeblastdb;4"
```

Looking at the HTML file `tool_test_output.html` via the `All tool test results.zip` available from the artefacts of the failed run:

```
Error: Unknown argument: "dbtypne"
Error:  (CArgException::eInvalidArg) Unknown argument: "dbtypne"
```

Happily that looks like an easy fix

**Review by wm75 — APPROVED** [review]
Thanks @peterjc and @elischberg - this looks good to me then!

**Comment by peterjc** [comment]
Squash-and-merged, thank you both.

In theory this will auto-deploy to the Tool Shed... https://github.com/peterjc/galaxy_blast/actions/runs/8003981806

**Comment by peterjc** [comment]
Test Tool Shed failed - could be in flux?

Main Tool Shed also failed:

```
Could not update ncbi_blast_plus
Unexpected HTTP status code: 500: {"err_msg": "Metadata may have been defined for some items in revision 'b6893f57f8d8'.  Correct the following problems if necessary and reset metadata.<br/><b>ncbi_makeblastdb.xml<\/b> - This file refers to a file named <b>all_fasta.loc.sample<\/b>.  Upload a file named <b>all_fasta.loc.sample.sample<\/b> to the repository to correct this error.<br/>"}
Failed to update repository contents for repository 'ncbi_blast_plus' on the main Tool Shed.
Repository metadata updated successfully for repository 'ncbi_blast_plus' on the main Tool Shed.
```

Looks like we need to add ``all_fasta.loc.sample`` to https://github.com/peterjc/galaxy_blast/blob/master/tools/ncbi_blast_plus/.shed.yml

**Comment by peterjc** [comment]
Huh - the automated deployment did go though even without the ``all_fasta.loc.sample`` file, so there are actually two auto-deployments to the main Tool Shed.

Anyway, these changes are now live on the Tool Shed 🎉

**Comment by peterjc** [comment]
@nsoranzo thanks - I didn't double check the TTS status 👍

**Comment by wm75** [comment]
> Anyway, these changes are now live on the Tool Shed 🎉

... meaning the new version should be available on Galaxy Europe by Monday.

Thanks a lot!

**Pull request merged**